skxray/cdi.py

```diff
@@ -119,9 +125,339 @@ def convolution(array1, array2):
+    sw_threshold : float
+        threshold used in shrinkwrap method
+
+    Returns
+    -------
+    new_sup : array
+        updated sample support
+    s_index : array
+        index for sample area
+    s_out_index : array
+        index not for sample area
+    """
+
+    sample_obj = np.abs(sample_obj)
+    gauss_fun = gauss(sample_obj.shape, sw_sigma)
+    gauss_fun = gauss_fun / np.max(gauss_fun)
+
+    conv_fun = convolution(sample_obj, gauss_fun)
+
+    conv_max = np.max(conv_fun)
+
+    s_index = np.where(conv_fun >= (sw_threshold*conv_max))
+    s_out_index = np.where(conv_fun < (sw_threshold*conv_max))
+
+    new_sup = np.zeros_like(sample_obj)
+    new_sup[s_index] = 1
+
+    return new_sup, s_index, s_out_index
+
+
+def pi_support(sample_obj, index_v):
+    """
+    Define sample shape by cutting unnecessary values.
+
+    Parameters
+    ----------
+    sample_obj : array
+        sample data
+    index_v : array
+        index to define sample area
+
+    Returns
+    -------
+    sample_obj : array
+        sample object with proper cut.
+    """
+    sample_obj = np.array(sample_obj)
+    sample_obj[index_v] = 0
+    return sample_obj
+
+
+def cal_relative_error(x_old, x_new):
+    """
+    Relative error is calculated as the ratio of the difference between the new and
+    the original arrays to the norm of the original array.
+
+    Parameters
+    ----------
+    x_old : array
+        previous data set
+    x_new : array
+        new data set
+
+    Returns
+    -------
+    float :
+        relative error
+    """
+    return np.linalg.norm(x_new - x_old)/np.linalg.norm(x_old)
+
+
+def cal_diff_error(sample_obj, diff_array):
+    """
+    Calculate the error in q space.
+
+    Parameters
+    ----------
+    sample_obj : array
+        sample data
+    diff_array : array
+        diffraction pattern
+
+    Returns
+    -------
+    float :
+        relative error in q space
+    """
+    new_diff = np.abs(_fft_helper(sample_obj)) / np.sqrt(np.size(sample_obj))
+    return cal_relative_error(diff_array, new_diff)
+
+
+def generate_random_phase_field(diff_array):
+    """
+    Initiate random phase.
+
+    Parameters
+    ----------
+    diff_array : array
+        diffraction pattern
+
+    Returns
+    -------
+    sample_obj : array
+        sample information with phase
+    """
+    pha_tmp = np.random.uniform(0, 2*np.pi, diff_array.shape)
+    sample_obj = (_ifft_helper(diff_array * np.exp(1j*pha_tmp))
+                  * np.sqrt(np.size(diff_array)))
+    return sample_obj
+
+
+def generate_box_support(sup_radius, shape_v):
+    """
+    Generate support area as a box for either 2D or 3D cases.
+
+    Parameters
+    ----------
+    sup_radius : float
+        radius of support
+    shape_v : list
+        shape of diffraction pattern, which can be either 2D or 3D case.
+
+    Returns
+    -------
+    sup : array
+        support with a box area
+    """
+    sup = np.zeros(shape_v)
```

Comment:
This can be done much more simply:

```python
slc_list = [slice(s//2 - sup_radius, s//2 + sup_radius) for s in shape_v]
sup = np.zeros(shapev)
sup[slc_lst] = 1
```

It is also better to be explicit about dealing with the case of one of the dimensions being odd rather than relying on numpy to do the right thing with slicing with floats (it seems to cast the values to ints so this is the same behavior).